simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java

```diff
@@ -0,0 +1,227 @@
+            }
+            policyStats.recordWeightedMiss(event.weight());
+            for (int i = 0; i < nW; i++) {
+                if (Math.round(event.missPenalty() / Math.pow(nQ, i)) > 0) {
+                    w = i;
+                }
+            }
+            int q = (int) (Math.round(event.missPenalty() / Math.pow(nQ, w)) + CH[w]) % nQ;
+            wheels[w].add(q, event);
+            currentSize += event.weight();
+            data.put(key, new Node(event, w, q));
+        } else {
+            policyStats.recordWeightedHit(event.weight());
+            wheels[node.wheel].remove(node.q, node.event);
+            for (int i = 0; i < nW; i++) {
+                if (Math.round(node.event.missPenalty() / Math.pow(nQ, i)) > 0) {
+                    w = i;
+                }
+            }
+            int q = (int) (Math.round(event.missPenalty() / Math.pow(nQ, w)) + CH[w]) % nQ;
+            node.updateLoc(w, q);
+            wheels[w].add(q, node.event);
+        }
+
+    }
+
+    private void migration(int idx) {
+        CH[idx] = (CH[idx] + 1) % nQ;
+        if (CH[idx] == 0 && idx + 1 < nW) {
+            migration(idx + 1);
+        }
+        while (wheels[idx].queueNotEmpty(CH[idx])) {
+            AccessEvent p = wheels[idx].evict(CH[idx]);
+            int cr = ((int) p.missPenalty()) % ((int) Math.pow(nQ, idx));
+            int q = (int) (Math.round(cr / Math.pow(nQ, idx - 1)) + CH[idx - 1]) % nQ;
+            wheels[idx - 1].add(q, p);
+            Node node = data.get(p.key());
+            node.updateLoc(idx - 1, q);
+        }
+    }
+
+    @Override
+    public PolicyStats stats() {
+        return policyStats;
+    }
+
+    static class GDWheelSettings extends BasicSettings {
+        public GDWheelSettings(Config config) {
+            super(config);
+        }
+
+        public int numberOfWheels() {
+            return config().getInt("gd-wheel.nw");
+        }
+
+        public int numberOfQueues() {
+            return config().getInt("gd-wheel.nq");
+        }
+    }
+
+    static class CostWheel {
+        List<TreeSet<AccessEvent>> wheel;
+        int current_index;
+        int nQ;
+
+        public CostWheel(int nQ) {
+            this.wheel = new ArrayList<>();
+            for (int i = 0; i < nQ; i++) {
+                this.wheel.add(new TreeSet<>(Comparator.comparingDouble(AccessEvent::missPenalty)));
+            }
+            this.current_index = 0;
+            this.nQ = nQ;
+        }
+
+        public int getNextIndex() {
+            for (int i = current_index; i < current_index + nQ; i++) {
+                int j = i % nQ;
+                if (!wheel.get(j).isEmpty()) {
+                    current_index = j;
+                    break;
+                }
+            }
+            return current_index;
+        }
+
+        public AccessEvent evict(int i) {
+            TreeSet<AccessEvent> queue = wheel.get(i);
+            return Objects.requireNonNull(queue.first());
+        }
```

Comment:
This is the bug that causes the cache to not evict correctly. The method`first()` peeks into the queue, whereas you want `pollFirst()` to remove it.

Reply:
This fix only works if always allowing admission, as if `admit = false` then it seems to break...